### DISCUSSION TITLE
Decimal value on teeth control on ABS/ESP insteed integer

### DISCUSSION THREAD

**Discussion by CedricM3RS** [cedricp/ddt4all#1233] [Q&A]
Hello,

My test is on a X91 ECU "ESP", version 4.0.7. 
I try to use the 3-33 panel teeth control to see the number of pulse on sensor magnet of ABS/ESP
A disc magnet of ABS sensor is a series of pair of south/north magnetic pole, so they can be 2,4,6 and for better resolution 40,42,44... then the number of magnet is only integer and even.

When I use this fonctionnality (on each front wheel that I tested), the result number is decimal. For a 1 exact turn, I have 30.47 insteed 47 teeth like comments on the same panel. When I turn just a little more or a little less than one exact turn, I see 29.xx and 31.xx, so there is indeed a variation who follows the movement

I do not understand how to interpret this value (and that is not conform with comment)
Here, the mesured value is 30.47, and this is disturbing because the expected value is 47 (on comments), is it a coincidence ?

Thank you if you can help me
Regards
CM

**Comment by shrlnm** [comment]
Hello @CedricM3RS 
What exact xml file are you using?

**Comment by CedricM3RS** [comment]
Hello,

file is ESP_X91_Version_4.0.7_20130424T093135.xml
![ESP_X91](https://github.com/user-attachments/assets/a916d32a-52bd-4e04-a37e-996d9ebac1ce)

I search into file about counting. I Find "PolesCounting_RR Start.Routine" request that correspond to 3-33 tab.
I try to search about counting function, I do not find any library on fille tree that have name like "couting".

Regards
CM

**Comment by CedricM3RS** [comment]
Hello
If it can help, I made a physical count of the magnetic target.
There are 96 poles, so 48 pairs (see attached)
The measurement with the software indicates exactly 30.47 counts for very precisely one complete turn.
With a stationary measurement, the software returns the value -0.25 instead of 0.
I had popup information of init counts and result
Regards
CM

![result_infos](https://github.com/user-attachments/assets/0227a33d-e91d-4412-adc8-2fa4d275edbf)

![init_counts_infos](https://github.com/user-attachments/assets/89ea2b31-9871-47e3-9813-014b7276ee16)

![nb_poles](https://github.com/user-attachments/assets/235d9fe3-3647-403a-a37b-073e9ee2735a)

**Comment by shrlnm** [comment]
@CedricM3RS 
The value displayed in this field is calculated using the following formula:

```
<Data Name="Nombre de dents">
<Bits>
<Scaled Step=" 64" Offset="-50" DivideBy=" 200" Format="##"/>
</Bits>


```
I.e. DisplayedValue = (x * 64-50)/200
So if you count it back   (30.47 * 200 + 50)/64 = 96

I do not know why they use such a formula here. But I think if you replace it  on the next one, you'll just get the number of poles.

<Scaled Step="1" Offset="0" DivideBy="1"/>

**Comment by CedricM3RS** [comment]
Hello,
Thanks a lot. This formula clearly explains the values (ie -0.25 for 0 and 30.47 for 96).

Indeed this formula is disturbing.
I search for other ECU to compare, and for X95 formula is
![formula_X95](https://github.com/user-attachments/assets/552ade7b-2bd0-43c7-ad1f-9117b88dbb08)
Then, good formula exists...

I searched through the X95 ECU versions because I initially expected this version, but the auto-detect selected X91, and the manual selection is no better.
X95 versions are older than X91, so I wonder if:
- either X91 ECUs replace X95s from a certain year, and the xml file does not take into account the variants,
- or the ECU is misdetected, in which case X91 is loaded instead of X95

I will try to do some tests to identify the initial problem (version? detection?), but in any case you have identified the calculation problem.
Thanks !

CM
